crates/sui-types/src/temporary_store.rs

```diff
@@ -435,33 +435,33 @@ impl<S> TemporaryStore<S> {
     ) -> Result<(), ExecutionError> {
         let mut objects_to_update = vec![];
         // Also charge gas for mutating the gas object in advance.
-        let gas_object_size = gas_object.object_size_for_gas_metering();
-        gas_object.storage_rebate = gas_status.charge_storage_mutation(
-            gas_object_size,
-            gas_object_size,
-            gas_object.storage_rebate.into(),
-        )?;
+        gas_object.charge_storage_mutation_no_size_change(gas_status)?;
         objects_to_update.push((
             SingleTxContext::gas(sender),
             gas_object.clone(),
             WriteKind::Mutate,
         ));
 
         for (object_id, (ctx, object, write_kind)) in &mut self.written {
-            let (old_object_size, storage_rebate) = self
-                .input_objects
-                .get(object_id)
-                .map(|old| (old.object_size_for_gas_metering(), old.storage_rebate))
-                .unwrap_or((0, 0));
-            let new_storage_rebate = gas_status.charge_storage_mutation(
-                old_object_size,
-                object.object_size_for_gas_metering(),
-                storage_rebate.into(),
-            )?;
+            match *write_kind {
+                WriteKind::Create => {
+                    object.charge_storage_mutation_no_size_change(gas_status)?
+                }
+                WriteKind::Mutate => {
+                    // TODO: what if object_id is a dynamic field or was unwrapped?
+                    let old_object =self
+                    .input_objects
+                    .get(object_id)
+                    .expect("Mutated an object, but there's no corresponding input. Dynamic field?");
+                    object.charge_storage_mutation(old_object, gas_status)?;
+                }
+                WriteKind::Unwrap => {
+                    // TODO: what happens if the object got unwrapped, then mutated? Will it show up as a mutation (ok), or
```

Comment:
Also not sure about how this will show up.

Reply:
Looking at the code a bit, I don't see a place in the code where create an instance of `WriteKind::Unwrap`--does that seem right @tnowacki @amnn ?

Reply:
The instance is created here: https://github.com/MystenLabs/sui/blob/main/crates/sui-types/src/messages.rs#L1791

Reply:
Unwrap is determined here
https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/src/natives/object_runtime/mod.rs#L320

Reply:
>  // TODO: what happens if the object got unwrapped, then mutated? Will it show up as a mutation (ok), or

But to answer your question, `Unwrap` is given when both 
- The object was not present in storage before the transaction (not an input, not a loaded child)
- The object ID was not created in this transaction.

Reply:
Thanks! I think this case is fine then, because a wrapped object will not have a storage rebate. Will combine with the `Create` case.